src/App.tsx

```diff
@@ -1,26 +1,170 @@
+    return todosFromServer.filter(todo => todo.completed).map(todo => todo.id);
+  }, [todosFromServer]);
+
+  const uncompletedTodos = useMemo(() => {
+    return todosFromServer.filter(todo => !todo.completed);
+  }, [todosFromServer]);
+
+  const uncompletedTodosAmount = uncompletedTodos.length;
+
+  const filteredTodos = useMemo(() => {
+    return getFilteredTodos(todosFromServer, filterOption);
+  }, [todosFromServer, filterOption]);
+
+  const handleDeleteTodos = useCallback((ids: number[]) => {
+    Promise.all(
+      ids.map(id => {
+        setDeletionIds(current => [...current, id]);
+
+        deleteTodo(id)
+          .then(() => {
+            setTodosFromServer(current =>
+              current.filter(todo => todo.id !== id),
+            );
+          })
+          .catch(() => {
+            handleError(setErrorMessage, Errors.DeleteTodo);
+          })
+          .finally(() =>
+            setDeletionIds(currIds => currIds.filter(currId => currId !== id)),
+          );
+      }),
+    );
+  }, []);
+
+  const handleChangeTodos = useCallback(
+    (newTodos: Todo[]) => {
+      return Promise.all(
+        newTodos.map(todo => {
+          setPendingTodos(current => [...current, todo]);
+
+          const { id, ...todoBody } = todo;
+
+          return patchTodo(todoBody, id)
+            .then(() => {
+              setTodosFromServer(current =>
+                current.map(currentTodo => {
+                  return currentTodo.id !== todo.id ? currentTodo : todo;
+                }),
+              );
+            })
+            .catch(() => {
+              handleError(setErrorMessage, Errors.UpdateTodo);
+              throw new Error(errorMessage);
+            })
+            .finally(() => setPendingTodos([]));
+        }),
+      );
+    },
+    [errorMessage],
+  );
+
+  const handleToggleCompleted = (id: number, updatedField: Partial<Todo>) => {
+    setPendingTodos(current => [...current]);
+
+    return patchTodo(updatedField, id)
+      .then((updatedTodo: Todo) =>
+        setTodosFromServer(current =>
+          current.map(todo => (todo.id === id ? updatedTodo : todo)),
+        ),
+      )
+      .catch(() => {
+        handleError(setErrorMessage, Errors.UpdateTodo);
+        throw new Error(errorMessage);
+      })
+      .finally(() =>
+        setPendingTodos(current => current.filter(todo => todo.id !== id)),
+      );
+  };
+
+  const handleToggleCompletedAll = () => {
+    if (uncompletedTodos && uncompletedTodosAmount) {
+      Promise.allSettled(
+        uncompletedTodos.map(({ id }) =>
+          handleToggleCompleted(id, { completed: true }),
+        ),
+      );
+    } else {
+      Promise.allSettled(
+        todosFromServer.map(({ id }) =>
+          handleToggleCompleted(id, { completed: false }),
+        ),
+      );
+    }
+  };
+
+  useEffect(() => {
+    getTodos()
+      .then(setTodosFromServer)
+      .catch(() => handleError(setErrorMessage, Errors.LoadingTodos));
```

Comment:
It's better to move the logic inside the `useEffect `hook to a separate function, as it improves code readability, reusability, and makes testing easier. It also helps keep `useEffect `focused on side effects.